# Sprint 1
# DATA Cleaning

In [1]:
import numpy as np
import pandas as pd

#### Loading the patients data into jupyter Noteook

In [2]:
# load the dataset's
patients = pd.read_csv('/content/patients.csv')

In [3]:
# viewing the data set patients
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [4]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [5]:
# finding the null values in each field
patients.isna().sum()

patient_id       0
assigned_sex     0
given_name       0
surname          0
address         12
city            12
state           12
zip_code        12
country         12
contact         12
birthdate        0
weight           0
height           0
bmi              0
dtype: int64

In [6]:
patients.shape

(503, 14)

### The Following Errors are observed in the Patients Dataset


##### 1. Address, City, State, Country, Zip_code columns have null values.
##### 2. Zip_code Datatype is in Float, it should be in Int.
##### 3. Contact Column has mixed informatation of email and mobile numbers of customers.
##### 4. The names in given_name and sur_name columns are containing special characters.
##### 5.The Birthdate column datatype is mismatched, the birthdate column is object datatype,but it should be in datetime64.
##### 6.The names in given_name and sur_name columns are containing special characters.
##### 7.The values in weight column is mentioned in pound, but they should be in kg's.
##### 8.The Values in Height columns are mentioned in inches, but they should be in meters's.
##### 9.The bmi column values are incorrectly calculated.
##### 10.The State Column has the Statecode insted the state column should contain the full name of the state



In [7]:
# fill the null values
patients.address.fillna('unknown',inplace = True)
patients.city.fillna('unknown',inplace = True)
patients.state.fillna('unknown',inplace = True)
patients.country.fillna('unknown',inplace = True)

# find and fill null values of zipcode
patients[patients['zip_code'].isna()]
patients.zip_code.fillna(0,inplace = True)

<ipython-input-7-71e651c02889>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  patients.address.fillna('unknown',inplace = True)
<ipython-input-7-71e651c02889>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using

In [8]:
# verifying the null values in zip_code
patients.zip_code.isna().sum()

0

#### Zip_code Datatype is in Float, it should be in Int

In [9]:
# converting the zip_code datatype from float to int
patients.zip_code = patients.zip_code.astype(int)

#### Contact Column has mixed informatation of email and mobile numbers of customers,so i am trying to separate the email and mobile numbers of customers and fill them into a new colums for each  

In [10]:
# extracting the email and phone number from contact
phone_pattern = r'(\+?\d{1,4}[\s-]?\(?\d{1,4}\)?[\s-]?\d{1,4}[\s-]?\d{1,9})'

patients['phone_number'] = patients['contact'].str.extract(phone_pattern)
patients['phone_number'] = patients['phone_number'].str.replace('[\(\)\+\-\s]', '', regex=True)

# filling the nan values with 0
patients['phone_number'].fillna('0', inplace=True)

#converting datatype of the phone_number to str
patients['phone_number'] = patients['phone_number'].astype(str)

#removing the countrycode and other digits from phone_number
def phone_number(number):
    if len(number) == 11:
        number = number[1:]
    return number

patients['phone_number'] = patients['phone_number'].apply(phone_number)
patients['phone_number'].head()


<ipython-input-10-da4b01ed3b49>:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  patients['phone_number'].fillna('0', inplace=True)


0    9517199170
1    2175693204
2    4023636804
3    7326368246
4    3345157487
Name: phone_number, dtype: object

In [11]:
# extracting the email from contact
email_pattern = r'([a-zA-Z._%+-]+@[a-zA-Z.-]+\.[a-zA-Z]{2,})'
patients['email'] = patients['contact'].str.extract(email_pattern)
# filling null values with Unknown
patients['email'].fillna('unknown',inplace = True)

<ipython-input-11-16d33471b42e>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  patients['email'].fillna('unknown',inplace = True)


#### Verifying the the email and mobile numbers whether they are separated

In [12]:
patients[['contact', 'email', 'phone_number']].head(20)


contact  \
0            951-719-9170ZoeWellish@superrito.com   
1           PamelaSHill@cuvox.de+1 (217) 569-3204   
2                402-363-6804JaeMDebord@gustr.com   
3      PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4                 334-515-7487TimNeudorf@cuvox.de   
5        386-334-5237RafaelCardosoCosta@gustr.com   
6               775-533-5933MaryBAdams@einrot.com   
7            XiuxiuChang@einrot.com1 408 778 3236   
8         816-265-9578DavidGustafsson@armyspy.com   
9   SophieCabreraIbarra@teleworm.us1 718 795 9124   
10         906-478-8949SandyGunnarsson@dayrep.com   
11     Abdul-NurMummarIsa@rhyta.com1 931 207 0839   
12      OmeokachieIbeamaka@einrot.com434-509-2614   
13     AnenechiChidi@armyspy.com+1 (205) 417-8095   
14              AsiaWozniak@rhyta.com918-712-3469   
15                276-225-1955SrenFLund@gustr.com   
16         LieuThiThuTam@dayrep.com1 559 765 7836   
17       RoxanneAndreyeva@armyspy.com678-829-8578   
18          406-775-2696WilliamVOates@armyspy.com   
19               ZakKelly@rhyta.com1 530 532 8397   

                              email phone_number  
0          ZoeWellish@superrito.com   9517199170  
1              PamelaSHill@cuvox.de   2175693204  
2              JaeMDebord@gustr.com   4023636804  
3         PhanBaLiem@jourrapide.com   7326368246  
4               TimNeudorf@cuvox.de   3345157487  
5      RafaelCardosoCosta@gustr.com   3863345237  
6             MaryBAdams@einrot.com   7755335933  
7            XiuxiuChang@einrot.com   4087783236  
8       DavidGustafsson@armyspy.com   8162659578  
9   SophieCabreraIbarra@teleworm.us   7187959124  
10       SandyGunnarsson@dayrep.com   9064788949  
11     Abdul-NurMummarIsa@rhyta.com   9312070839  
12    OmeokachieIbeamaka@einrot.com   4345092614  
13        AnenechiChidi@armyspy.com   2054178095  
14            AsiaWozniak@rhyta.com   9187123469  
15              SrenFLund@gustr.com   2762251955  
16         LieuThiThuTam@dayrep.com   5597657836  
17     RoxanneAndreyeva@armyspy.com   6788298578  
18        WilliamVOates@armyspy.com   4067752696  
19               ZakKelly@rhyta.com   5305328397

In [13]:
# removing the patients contact column
patients.drop('contact',axis = 1, inplace = True)

In [14]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California     92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois     61812  United States   4/3/1967   118.8   
2               York    Nebraska     68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ      7095  United States  7/26/1951   220.9   
4             Dothan          AL     36303  United States  2/18/1928   192.3   

   height   bmi phone_number                      email  
0      66  19.6   9517199170   ZoeWellish@superrito.com  
1      66  19.2   2175693204       PamelaSHill@cuvox.de  
2      71  24.8   4023636804       JaeMDebord@gustr.com  
3      70  31.7   7326368246  PhanBaLiem@jourrapide.com  
4      27  26.1   3345157487        TimNeudorf@cuvox.de

#### The Birthdate column datatype is mismatched, the birthdate column is object datatype,but it should be in datetime64.

In [15]:
# changinng the birthdate field datatype from object to datetime64
patients['birthdate'] = pd.to_datetime(patients['birthdate'])

In [16]:
patients.birthdate.info()

<class 'pandas.core.series.Series'>
RangeIndex: 503 entries, 0 to 502
Series name: birthdate
Non-Null Count  Dtype         
--------------  -----         
503 non-null    datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 4.1 KB


### The names in given_name and sur_name columns are containing special characters, so i am trying to replace them with the unicode characters.

In [17]:
# cleaning the errors in names
patients[['given_name','surname']].head(10)

given_name     surname
0        Zoe     Wellish
1     Pamela        Hill
2        Jae      Debord
3       Liêm        Phan
4        Tim     Neudorf
5     Rafael       Costa
6       Mary       Adams
7     Xiuxiu       Chang
8      Dsvid  Gustafsson
9     Sophie     Cabrera

In [18]:
pip install unidecode

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.5/235.5 kB 8.2 MB/s eta 0:00:00


In [19]:
import unidecode

# converting the the special character in names to regular characters
def normalize_name(name):
    return unidecode.unidecode(name)

# Apply the function to the given_name & sur_name columns
patients['given_name'] = patients['given_name'].apply(normalize_name)
patients['surname'] = patients['surname'].apply(normalize_name)
patients[['given_name','surname']].head(10)


given_name     surname
0        Zoe     Wellish
1     Pamela        Hill
2        Jae      Debord
3       Liem        Phan
4        Tim     Neudorf
5     Rafael       Costa
6       Mary       Adams
7     Xiuxiu       Chang
8      Dsvid  Gustafsson
9     Sophie     Cabrera

#### Creating a new column for Full_name of the customers

In [20]:
# creating Full name column
patients['full_name'] = patients['given_name'] + ' ' + patients['surname']
patients['full_name'].head()

0    Zoe Wellish
1    Pamela Hill
2     Jae Debord
3      Liem Phan
4    Tim Neudorf
Name: full_name, dtype: object

#### The values in weight column is mentioned in pound, but they should be in kg's
#### The Values in Height columns are mentioned in inches, but they should be in meters's

In [21]:
#converting the weight from pounds to kg's
patients['weight'] = patients['weight'] * 0.453592

# converting the height from inches to mt's
patients['height'] = patients['height'] * 0.0254

#### The bmi column values are incorrectly calculated,so i am trying to calculate the correct bmi with revised weight's and Height's of the customer's
##### bmi: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m2 where kg is a person's weight in kilograms and m2 is their height in metres squared.

In [22]:
#BMI = kg/m2 calculation and creating column
patients['bmi'] = patients['weight'] / (patients['height'] **2)

In [23]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liem         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindstrom   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California     92390   
1    2370 University Hill Road          Armstrong    Illinois     61812   
2        1493 Poling Farm Road               York    Nebraska     68467   
3          2335 Webster Street         Woodbridge          NJ      7095   
4         1428 Turkey Pen Lane             Dothan          AL     36303   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME      3852   
499   494 Clarksburg Park Road             Sedona          AZ     86341   
500          649 Nutter Street      Overland Park          MO     64110   
501   3652 Boone Crockett Lane            Seattle          WA     98109   
502          2778 North Avenue               Burr    Nebraska     68324   

           country  birthdate      weight  height         bmi phone_number  \
0    United States 1976-07-10   55.202146  1.6764   19.642676   9517199170   
1    United States 1967-04-03   53.886730  1.6764   19.174609   2175693204   
2    United States 1980-02-19   80.648658  1.8034   24.797792   4023636804   
3    United States 1951-07-26  100.198473  1.7780   31.695499   7326368246   
4    United States 1928-02-18   87.225742  0.6858  185.459767   3345157487   
..             ...        ...         ...     ...         ...          ...   
498  United States 1959-04-10   82.145511  1.8288   24.561303   2074770579   
499  United States 1948-03-26  108.680643  1.7780   34.378640   9282844492   
500  United States 1971-01-13   77.654950  1.7018   26.813414   8162236007   
501  United States 1952-02-13   80.240425  1.7018   27.706150   3604432060   
502  United States 1954-05-03   62.686414  1.8034   19.274774   4028484923   

                                 email            full_name  
0             ZoeWellish@superrito.com          Zoe Wellish  
1                 PamelaSHill@cuvox.de          Pamela Hill  
2                 JaeMDebord@gustr.com           Jae Debord  
3            PhanBaLiem@jourrapide.com            Liem Phan  
4                  TimNeudorf@cuvox.de          Tim Neudorf  
..                                 ...                  ...  
498    MustafaLindstrom@jourrapide.com    Mustafa Lindstrom  
499             RumanBisliev@gustr.com        Ruman Bisliev  
500          JinkedeKeizer@teleworm.us      Jinke de Keizer  
501  ChidaluOnyekaozulu@jourrapide.com  Chidalu Onyekaozulu  
502           PatrickGersten@rhyta.com          Pat Gersten  

[503 rows x 16 columns]

#### Calculating the bmi_status using the bmi values
##### A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. The inclusion criteria for this clinical trial is 16 >= BMI >= 38.


In [24]:
def status(bmi):
  if bmi < 18.5:
      return 'Underweight'
  elif 18.5 <= bmi < 24.9:
      return 'Normal weight'
  elif 24.9 <= bmi < 29.9:
      return 'Overweight'
  elif 29.9 <= bmi < 34.9:
      return 'Obesity'
  else:
      return 'Severe Obesity'

# Apply the function to each value in the 'revised_bmi' column
Bmi_Status = patients['bmi'].apply(status)


In [25]:
# creating the Bmi_status column
patients['bmi_status'] = Bmi_Status
patients['bmi_status']

0       Normal weight
1       Normal weight
2       Normal weight
3             Obesity
4      Severe Obesity
            ...      
498     Normal weight
499           Obesity
500        Overweight
501        Overweight
502     Normal weight
Name: bmi_status, Length: 503, dtype: object

#### The State Column has the Statecode insted the state column should contain the full name of the state

In [26]:
patients.state.unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       'unknown', 'VT', 'ID', 'DC', 'AR'], dtype=object)

In [27]:
def full_name(state):
    states_full_name = {
        'AL': 'Alabama', 'AK': 'Alaska', 'AZ': 'Arizona', 'AR': 'Arkansas', 'CA': 'California',
        'CO': 'Colorado', 'CT': 'Connecticut', 'DE': 'Delaware', 'FL': 'Florida', 'GA': 'Georgia',
        'HI': 'Hawaii', 'ID': 'Idaho', 'IL': 'Illinois', 'IN': 'Indiana', 'IA': 'Iowa',
        'KS': 'Kansas', 'KY': 'Kentucky', 'LA': 'Louisiana', 'ME': 'Maine', 'MD': 'Maryland',
        'MA': 'Massachusetts', 'MI': 'Michigan', 'MN': 'Minnesota', 'MS': 'Mississippi',
        'MO': 'Missouri', 'MT': 'Montana', 'NE': 'Nebraska', 'NV': 'Nevada', 'NH': 'New Hampshire',
        'NJ': 'New Jersey', 'NM': 'New Mexico', 'NY': 'New York', 'NC': 'North Carolina',
        'ND': 'North Dakota', 'OH': 'Ohio', 'OK': 'Oklahoma', 'OR': 'Oregon', 'PA': 'Pennsylvania',
        'RI': 'Rhode Island', 'SC': 'South Carolina', 'SD': 'South Dakota', 'TN': 'Tennessee',
        'TX': 'Texas', 'UT': 'Utah', 'VT': 'Vermont', 'VA': 'Virginia', 'WA': 'Washington',
        'WV': 'West Virginia', 'WI': 'Wisconsin', 'WY': 'Wyoming', 'DC': 'District of Columbia'
    }
    return states_full_name.get(state, state)

# Applying the function to the 'state' column to convert the shortform's to full name
patients['state'] = patients['state'].apply(full_name)

# Displaying the updated 'state' column
patients['state']


0      California
1        Illinois
2        Nebraska
3      New Jersey
4         Alabama
          ...    
498         Maine
499       Arizona
500      Missouri
501    Washington
502      Nebraska
Name: state, Length: 503, dtype: object

In [28]:
#finding repeted / Duplicate entries
patients['full_name'].value_counts().nlargest()

full_name
John Doe           6
Zoe Wellish        1
Jad Allah Gaber    1
Asuna Morita       1
Sophia Haugen      1
Name: count, dtype: int64

In [29]:
patients.loc[patients['full_name'] == 'John Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

        state  zip_code        country  birthdate    weight  height  \
215  New York     12345  United States 1975-01-01  81.64656  1.8288   
229  New York     12345  United States 1975-01-01  81.64656  1.8288   
237  New York     12345  United States 1975-01-01  81.64656  1.8288   
244  New York     12345  United States 1975-01-01  81.64656  1.8288   
251  New York     12345  United States 1975-01-01  81.64656  1.8288   
277  New York     12345  United States 1975-01-01  81.64656  1.8288   

           bmi phone_number              email full_name     bmi_status  
215  24.412118   1234567890  johndoe@email.com  John Doe  Normal weight  
229  24.412118   1234567890  johndoe@email.com  John Doe  Normal weight  
237  24.412118   1234567890  johndoe@email.com  John Doe  Normal weight  
244  24.412118   1234567890  johndoe@email.com  John Doe  Normal weight  
251  24.412118   1234567890  johndoe@email.com  John Doe  Normal weight  
277  24.412118   1234567890  johndoe@email.com  John Doe  Normal weight

In [30]:
# removing duplicates
pat_id = [230, 238, 245, 252, 278]
patients = patients[~patients['patient_id'].isin(pat_id)]


In [31]:
patients['full_name'].value_counts().nlargest()

full_name
Zoe Wellish      1
Abel Efrem       1
Asuna Morita     1
Sophia Haugen    1
Emy Markus       1
Name: count, dtype: int64

#### Creating age column
###### Assuming the year of data collection is 2019

In [32]:
patients['age'] = 2019 - patients['birthdate'].dt.year

<ipython-input-32-fbcf9eba99a4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  patients['age'] = 2019 - patients['birthdate'].dt.year


In [33]:
patients.shape

(498, 18)

In [34]:
# re-arranging the column's order
column_order = [
    'patient_id', 'given_name', 'surname', 'full_name','assigned_sex', 'weight',
    'height','birthdate', 'age', 'phone_number', 'email',  'address',
    'city', 'state', 'zip_code', 'country', 'bmi', 'bmi_status', ]
patients = patients[column_order]

In [35]:
patients.head()

patient_id given_name  surname    full_name assigned_sex      weight  \
0           1        Zoe  Wellish  Zoe Wellish       female   55.202146   
1           2     Pamela     Hill  Pamela Hill       female   53.886730   
2           3        Jae   Debord   Jae Debord         male   80.648658   
3           4       Liem     Phan    Liem Phan         male  100.198473   
4           5        Tim  Neudorf  Tim Neudorf         male   87.225742   

   height  birthdate  age phone_number                      email  \
0  1.6764 1976-07-10   43   9517199170   ZoeWellish@superrito.com   
1  1.6764 1967-04-03   52   2175693204       PamelaSHill@cuvox.de   
2  1.8034 1980-02-19   39   4023636804       JaeMDebord@gustr.com   
3  1.7780 1951-07-26   68   7326368246  PhanBaLiem@jourrapide.com   
4  0.6858 1928-02-18   91   3345157487        TimNeudorf@cuvox.de   

                     address               city       state  zip_code  \
0       576 Brown Bear Drive  Rancho California  California     92390   
1  2370 University Hill Road          Armstrong    Illinois     61812   
2      1493 Poling Farm Road               York    Nebraska     68467   
3        2335 Webster Street         Woodbridge  New Jersey      7095   
4       1428 Turkey Pen Lane             Dothan     Alabama     36303   

         country         bmi      bmi_status  
0  United States   19.642676   Normal weight  
1  United States   19.174609   Normal weight  
2  United States   24.797792   Normal weight  
3  United States   31.695499         Obesity  
4  United States  185.459767  Severe Obesity

In [36]:
patients.to_csv('patients_cleaned.csv', index=False)

# adverse reactions

In [37]:
# Load the data
reaction = pd.read_csv('/content/adverse_reactions.csv')

In [38]:
reaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [39]:
reaction.isnull().sum()

given_name          0
surname             0
adverse_reaction    0
dtype: int64

In [40]:
reaction['full_name'] = reaction['given_name'] + ' ' + reaction['surname']
reaction['full_name']

0         berta napolitani
1                lena baer
2               joseph day
3        flavia fiorentino
4          manouck wubbels
5            jasmine sykes
6           louise johnson
7          albinca komavec
8               noe aranda
9          sofia hermansen
10           tegan johnson
11            abel yonatan
12           abdul-nur isa
13             leon scholz
14        gabriele saenger
15             jia li teng
16          jakob jakobsen
17    christopher woodward
18            ole petersen
19         finley chandler
20          anenechi chidi
21     miłosław wiśniewski
22             lixue hsueh
23            merci leroux
24                kang mai
25       elliot richardson
26          clinton miller
27            idalia moore
28            xiuxiu chang
29           alex crawford
30           monika lončar
31              steven roy
32          cecilie nilsen
33        krisztina magyar
Name: full_name, dtype: object

In [41]:
reaction.full_name.duplicated().sum()

0

In [42]:
column_order = ['given_name', 'surname', 'full_name', 'adverse_reaction']
reaction = reaction[column_order]

In [43]:
reaction.to_csv("adverse_reactions_cleaned.csv",index = False)

# Treatments

In [44]:
# Load the data
treatments = pd.read_csv('/content/treatments.csv')

In [45]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [46]:
treatments.isnull().sum()

given_name        0
surname           0
auralin           0
novodra           0
hba1c_start       0
hba1c_end         0
hba1c_change    109
dtype: int64

In [47]:
treatments.shape

(280, 7)

In [48]:
treatments[['given_name', 'surname', 'auralin', 'novodra', 'hba1c_start',
       'hba1c_end', 'hba1c_change']].duplicated().sum()

1

# Following are the errors in Treatments Dataset that should be cleaned.

##### 1.The names in given_name and sur_name columns are containing special characters and '-' in characters.
##### 2. 1 duplicate value is present in data.
##### 3. hba1c column has 109 null values.
##### 4. auralin and novodra column values are mised with the 'u' units,it should be separated and re calculated.
##### 5.hba1c column has incorrect values it should be re calculated.

In [49]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [50]:
treatments.given_name.unique()

array(['veronika', 'elliot', 'yukitaka', 'skye', 'alissa', 'jasmine',
       'sophia', 'eddie', 'saber', 'asia', 'joseph', 'kristiina',
       'roxanne', 'gregor', 'simone', 'enco', 'camilla', 'gina',
       'addolorata', 'khalid', 'mile', 'tekla', 'brancaleone', 'chiemela',
       'isac', 'benoît', 'suhaim', 'mizuki', 'clinton', 'eugene',
       'szilveszter', 'bao', 'laura', 'cannan', 'alexander', 'csaba',
       'chỉ', 'władysław', 'kristján', 'krisztina', 'ásta', 'mahmud',
       'marija', 'jens', 'sauli', 'zdeněk', 'mariana', 'jad allah', 'zak',
       'jackson', 'albinca', 'kristoffer', 'mỹ', 'nasser', 'oles',
       'triana.', 'juliusz', 'klementyna', 'gabryŝ', 'leixandre',
       'onyekachukwu', 'mijndert', 'hanka', 'coralie', 'cumelén',
       'elisabeth', 'louise', 'joe', 'fatimah', 'ivan', 'tosh', 'hugo',
       'gabrielle', 'mette', 'ole', 'mackenzie', 'noriyuki', 'placido',
       'julia', 'avdei', 'hideki', 'robert', 'maya', 'else', 'furuta',
       'nilton', 'ananías', '

In [51]:
treatments.surname.unique()

array(['jindrová', 'richardson', 'takenaka', 'gormanston', 'montez',
       'sykes', 'haugen', 'archer', 'ménard', 'woźniak', 'day', 'hyypiä',
       'andreyeva', 'bole', 'baumgaertner', 'žibrik', 'zaitseva', 'cain',
       'lombardi', 'johnsrud', 'stanić', 'walczak', 'russo', 'tobeolisa',
       'berg', 'bonami', 'rahal', 'iwata', 'miller', 'mironov', 'totth',
       'she', 'ehrlichmann', 'cabrera', 'mathiesen', 'sági', 'lâm',
       'wieczorek', 'ingason', 'magyar', 'grímsdóttir', 'kadyrov',
       'grubišić', 'poulsen', 'koivuniemi', 'synek', 'souza', 'gaber',
       'kelly', 'addison', 'komavec', 'martinsen', 'quynh', 'mansour',
       'zhdanov', 'terrazas', 'majewski', 'sokołowska', 'tomaszewski',
       'alanis', 'obinna', 'ouwendijk', 'gegič', 'allaire', 'miramontes',
       'dimmen', 'johnson', 'edwards', 'kinfe', 'fomin', 'jensen',
       'collins', 'bidwill', 'sandgreen', 'petersen', 'mckay', 'sakai',
       'udinesi', 'werner', 'tikhonov', 'haraguchi', 'wagner', 'isaksson',


#### The names in given_name and sur_name columns are containing special characters and '-' in characters, so i am trying to replace them with the unicode characters

In [52]:
import unidecode

# converting the the special character in names to regular characters
def normalize_name(name):
    return unidecode.unidecode(name)

# Apply the function to the given_name & sur_name columns
treatments['given_name'] = treatments['given_name'].apply(normalize_name)
treatments['surname'] = treatments['surname'].apply(normalize_name)
treatments['given_name'] = treatments['given_name'].str.replace('-','')
treatments['surname'] = treatments['surname'].str.replace('-','')


In [53]:
# verifying the corrected names
treatments['given_name'].unique()

array(['veronika', 'elliot', 'yukitaka', 'skye', 'alissa', 'jasmine',
       'sophia', 'eddie', 'saber', 'asia', 'joseph', 'kristiina',
       'roxanne', 'gregor', 'simone', 'enco', 'camilla', 'gina',
       'addolorata', 'khalid', 'mile', 'tekla', 'brancaleone', 'chiemela',
       'isac', 'benoit', 'suhaim', 'mizuki', 'clinton', 'eugene',
       'szilveszter', 'bao', 'laura', 'cannan', 'alexander', 'csaba',
       'chi', 'wladyslaw', 'kristjan', 'krisztina', 'asta', 'mahmud',
       'marija', 'jens', 'sauli', 'zdenek', 'mariana', 'jad allah', 'zak',
       'jackson', 'albinca', 'kristoffer', 'my', 'nasser', 'oles',
       'triana.', 'juliusz', 'klementyna', 'gabrys', 'leixandre',
       'onyekachukwu', 'mijndert', 'hanka', 'coralie', 'cumelen',
       'elisabeth', 'louise', 'joe', 'fatimah', 'ivan', 'tosh', 'hugo',
       'gabrielle', 'mette', 'ole', 'mackenzie', 'noriyuki', 'placido',
       'julia', 'avdei', 'hideki', 'robert', 'maya', 'else', 'furuta',
       'nilton', 'ananias', '

In [54]:
treatments['surname'].unique()

array(['jindrova', 'richardson', 'takenaka', 'gormanston', 'montez',
       'sykes', 'haugen', 'archer', 'menard', 'wozniak', 'day', 'hyypia',
       'andreyeva', 'bole', 'baumgaertner', 'zibrik', 'zaitseva', 'cain',
       'lombardi', 'johnsrud', 'stanic', 'walczak', 'russo', 'tobeolisa',
       'berg', 'bonami', 'rahal', 'iwata', 'miller', 'mironov', 'totth',
       'she', 'ehrlichmann', 'cabrera', 'mathiesen', 'sagi', 'lam',
       'wieczorek', 'ingason', 'magyar', 'grimsdottir', 'kadyrov',
       'grubisic', 'poulsen', 'koivuniemi', 'synek', 'souza', 'gaber',
       'kelly', 'addison', 'komavec', 'martinsen', 'quynh', 'mansour',
       'zhdanov', 'terrazas', 'majewski', 'sokolowska', 'tomaszewski',
       'alanis', 'obinna', 'ouwendijk', 'gegic', 'allaire', 'miramontes',
       'dimmen', 'johnson', 'edwards', 'kinfe', 'fomin', 'jensen',
       'collins', 'bidwill', 'sandgreen', 'petersen', 'mckay', 'sakai',
       'udinesi', 'werner', 'tikhonov', 'haraguchi', 'wagner', 'isaksson',


In [55]:
treatments['full_name'] = treatments['given_name'] + ' ' + treatments['surname']
treatments['full_name']

0        veronika jindrova
1        elliot richardson
2        yukitaka takenaka
3          skye gormanston
4            alissa montez
              ...         
275        albina zetticci
276       john teichelmann
277         mathea lillebo
278          vallie prince
279    samuel gudbrandsson
Name: full_name, Length: 280, dtype: object

In [56]:
treatments['full_name'].duplicated().sum()

1

In [57]:
treatments['full_name'].value_counts().nlargest()

full_name
joseph day                2
veronika jindrova         1
svanhvit gudjonsdottir    1
monika loncar             1
martha afanasyeva         1
Name: count, dtype: int64

In [58]:
treatments[treatments['full_name'] == 'joseph day']

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
10      joseph     day  29u - 36u       -          7.7       7.19   
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change   full_name  
10            NaN  joseph day  
136           NaN  joseph day

In [59]:
treatments = treatments.drop(136)


In [60]:
treatments[treatments['full_name'] == 'joseph day']

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
10     joseph     day  29u - 36u       -          7.7       7.19   

    hba1c_change   full_name  
10           NaN  joseph day

In [61]:
treatments.shape

(279, 8)

In [62]:
treatments.auralin

0      41u - 48u
1              -
2              -
3      33u - 36u
4              -
         ...    
275    45u - 51u
276            -
277    23u - 36u
278    31u - 38u
279    53u - 56u
Name: auralin, Length: 279, dtype: object

In [63]:
treatments.novodra

0              -
1      40u - 45u
2      39u - 36u
3              -
4      33u - 29u
         ...    
275            -
276    49u - 49u
277            -
278            -
279            -
Name: novodra, Length: 279, dtype: object

#### The Baseline and end dose of auralin and novodra are in units and the change in the dose from baseline(start) and end was not mentioned. here i'm' calculating the change for auraline and novodra dose's

In [64]:
# Extracting the start and end doses for Auralin
auralin_start = treatments['auralin'].str.extract(r'(\d+)u -')[0]
auralin_end = treatments['auralin'].str.extract(r'- (\d+)u')[0]

# Converting the extracted values to numeric
treatments['auralin_start'] = pd.to_numeric(auralin_start, errors='coerce')
treatments['auralin_end'] = pd.to_numeric(auralin_end, errors='coerce')


In [65]:
# Extracting the start and end doses for Novodra
novodra_start = treatments['novodra'].str.extract(r'(\d+)u -')[0]
novodra_end = treatments['novodra'].str.extract(r'- (\d+)u')[0]

# Converting the extracted values to numeric
treatments['novodra_start'] = pd.to_numeric(novodra_start, errors='coerce')
treatments['novodra_end'] = pd.to_numeric(novodra_end, errors='coerce')


In [66]:
treatments['start_dose'] = treatments['auralin_start'].combine_first(treatments['novodra_start'])
treatments['end_dose'] = treatments['auralin_end'].combine_first(treatments['novodra_end'])

In [67]:
treatments['drug_type'] = treatments.apply(lambda row: 'Auralin' if pd.notna(row['auralin_start']) else 'Novodra', axis=1)

In [68]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrova  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change          full_name  auralin_start  auralin_end  novodra_start  \
0           NaN  veronika jindrova           41.0         48.0            NaN   
1          0.97  elliot richardson            NaN          NaN           40.0   
2           NaN  yukitaka takenaka            NaN          NaN           39.0   
3          0.35    skye gormanston           33.0         36.0            NaN   
4          0.32      alissa montez            NaN          NaN           33.0   

   novodra_end  start_dose  end_dose drug_type  
0          NaN        41.0      48.0   Auralin  
1         45.0        40.0      45.0   Novodra  
2         36.0        39.0      36.0   Novodra  
3          NaN        33.0      36.0   Auralin  
4         29.0        33.0      29.0   Novodra

#### The hba1c_change column values are incorrectly calculated, so i'm re calculating the hba1c_change values.hba1c_change = hba1c_start - hba1c_end

In [69]:
treatments['hba1c_change'] = treatments['hba1c_start'] - treatments['hba1c_end']

In [70]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrova  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change          full_name  auralin_start  auralin_end  novodra_start  \
0          0.43  veronika jindrova           41.0         48.0            NaN   
1          0.47  elliot richardson            NaN          NaN           40.0   
2          0.43  yukitaka takenaka            NaN          NaN           39.0   
3          0.35    skye gormanston           33.0         36.0            NaN   
4          0.32      alissa montez            NaN          NaN           33.0   

   novodra_end  start_dose  end_dose drug_type  
0          NaN        41.0      48.0   Auralin  
1         45.0        40.0      45.0   Novodra  
2         36.0        39.0      36.0   Novodra  
3          NaN        33.0      36.0   Auralin  
4         29.0        33.0      29.0   Novodra

In [71]:
treatments.drop(['auralin','novodra'],axis = 1,inplace = True)

In [72]:
# univariate Analysis of auralin_change
treatments.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change  \
0   veronika    jindrova         7.63       7.20          0.43   
1     elliot  richardson         7.56       7.09          0.47   
2   yukitaka    takenaka         7.68       7.25          0.43   
3       skye  gormanston         7.97       7.62          0.35   
4     alissa      montez         7.78       7.46          0.32   

           full_name  auralin_start  auralin_end  novodra_start  novodra_end  \
0  veronika jindrova           41.0         48.0            NaN          NaN   
1  elliot richardson            NaN          NaN           40.0         45.0   
2  yukitaka takenaka            NaN          NaN           39.0         36.0   
3    skye gormanston           33.0         36.0            NaN          NaN   
4      alissa montez            NaN          NaN           33.0         29.0   

   start_dose  end_dose drug_type  
0        41.0      48.0   Auralin  
1        40.0      45.0   Novodra  
2        39.0      36.0   Novodra  
3        33.0      36.0   Auralin  
4        33.0      29.0   Novodra

In [73]:
column_order = ['given_name', 'surname', 'full_name', 'auralin_start', 'auralin_end', 'novodra_start',
       'novodra_end', 'start_dose', 'end_dose',  'hba1c_start', 'hba1c_end', 'hba1c_change','drug_type']
treatments = treatments[column_order]

In [74]:
treatments.head()

given_name     surname          full_name  auralin_start  auralin_end  \
0   veronika    jindrova  veronika jindrova           41.0         48.0   
1     elliot  richardson  elliot richardson            NaN          NaN   
2   yukitaka    takenaka  yukitaka takenaka            NaN          NaN   
3       skye  gormanston    skye gormanston           33.0         36.0   
4     alissa      montez      alissa montez            NaN          NaN   

   novodra_start  novodra_end  start_dose  end_dose  hba1c_start  hba1c_end  \
0            NaN          NaN        41.0      48.0         7.63       7.20   
1           40.0         45.0        40.0      45.0         7.56       7.09   
2           39.0         36.0        39.0      36.0         7.68       7.25   
3            NaN          NaN        33.0      36.0         7.97       7.62   
4           33.0         29.0        33.0      29.0         7.78       7.46   

   hba1c_change drug_type  
0          0.43   Auralin  
1          0.47   Novodra  
2          0.43   Novodra  
3          0.35   Auralin  
4          0.32   Novodra

In [75]:
treatments.drop(['auralin_start', 'auralin_end', 'novodra_start', 'novodra_end'], axis = 1, inplace = True)

<ipython-input-75-7cb9339b8bea>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments.drop(['auralin_start', 'auralin_end', 'novodra_start', 'novodra_end'], axis = 1, inplace = True)


In [76]:
treatments.to_csv("treatments_cleaned.csv",index = False)

# Treatments_cut

In [77]:
# Load the data
treatments_cut = pd.read_csv('/content/treatments_cut.csv')

In [78]:
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [79]:
treatments_cut.isnull().sum()

given_name       0
surname          0
auralin          0
novodra          0
hba1c_start      0
hba1c_end        0
hba1c_change    28
dtype: int64

In [80]:
treatments_cut.shape

(70, 7)

In [81]:
treatments_cut.duplicated().sum()

0

# The following are the Errors in Treatments_cut Dataset

##### 1.The names in given_name and sur_name columns are containing special characters.
##### 2.The Baseline and end dose of auralin and novodra are in units and the change in the dose from baseline(start) and end was not mentioned.

In [82]:
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [83]:
treatments_cut['given_name'].unique()

array(['jožka', 'inunnguaq', 'alwin', 'thể', 'amanda', 'meya', 'sara',
       'frydryk', 'jane', 'angela', 'edelma', 'gabriela', 'annika',
       'bernarda', 'chiho', 'dani', 'mijael', 'emma', 'flavia', 'noe',
       'seweryna', 'david', 'beatrycze', 'miłosław', 'lóa', 'siebrigje',
       'firenze', 'zoe', 'una', 'luboš', 'trifon', 'galit', 'kifle',
       'meaza', 'adlan', 'alvin', 'chibuzo', 'sofia', 'guðni', 'maryam',
       'eufemio', 'dalmacia', 'daimy', 'hans peter', 'jeremy',
       'štefanija', 'anja', 'regolo', 'nebechi', 'satsita', 'timothy',
       'bjørnar', 'vivien', 'genet', 'borna', 'mary', 'blanka', 'joseph',
       'christopher', 'robert', 'maret', 'ella', 'fakhri', 'žarka',
       'lixue', 'rovzan', 'jakob', 'bernd', 'berta', 'armina'],
      dtype=object)

In [84]:
treatments_cut['surname'].unique()

array(['resanovič', 'heilmann', 'svensson', 'lương', 'ribeiro', 'sjöberg',
       'miles', 'adamski', 'citizen', 'lavrentyev', 'villalpando',
       'michalčíková', 'vaara', 'cindrić', 'higa', 'antoun', 'guerra',
       'barlow', 'fiorentino', 'aranda', 'czerwinska', 'villadsen',
       'woźniak', 'wiśniewski', 'hrafnsdóttir', 'koldenhof', 'fodor',
       'wellish', 'traustadóttir', 'pecha', 'izmailov', 'casárez',
       'mustafa', 'brhane', 'shishani', 'jackson', 'okoli', 'hermansen',
       'heimisson', 'dratchev', 'rosario', 'madrid', 'tromp', 'radislav',
       'montagu', 'novosel', 'mueller', 'nucci', 'ekechukwu', 'batukayev',
       'cotton', 'nilsen', 'zsinkó', 'fesahaye', 'šlezinger', 'adams',
       'jurković', 'tucker', 'woodward', 'maslov', 'sultygov', 'lund',
       'fakhoury', 'rap', 'hsueh', 'kishiev', 'jakobsen', 'schneider',
       'napolitani', 'sauvé'], dtype=object)

#### The names in given_name and sur_name columns are containing special characters, so i am trying to replace them with the unicode characters

In [85]:
import unidecode

# converting the the special character in names to regular characters
def normalize_name(name):
    return unidecode.unidecode(name)

# Apply the function to the given_name & sur_name columns
treatments_cut['given_name'] = treatments_cut['given_name'].apply(normalize_name)
treatments_cut['surname'] = treatments_cut['surname'].apply(normalize_name)

treatments_cut['given_name'].unique()

array(['jozka', 'inunnguaq', 'alwin', 'the', 'amanda', 'meya', 'sara',
       'frydryk', 'jane', 'angela', 'edelma', 'gabriela', 'annika',
       'bernarda', 'chiho', 'dani', 'mijael', 'emma', 'flavia', 'noe',
       'seweryna', 'david', 'beatrycze', 'miloslaw', 'loa', 'siebrigje',
       'firenze', 'zoe', 'una', 'lubos', 'trifon', 'galit', 'kifle',
       'meaza', 'adlan', 'alvin', 'chibuzo', 'sofia', 'gudni', 'maryam',
       'eufemio', 'dalmacia', 'daimy', 'hans peter', 'jeremy',
       'stefanija', 'anja', 'regolo', 'nebechi', 'satsita', 'timothy',
       'bjornar', 'vivien', 'genet', 'borna', 'mary', 'blanka', 'joseph',
       'christopher', 'robert', 'maret', 'ella', 'fakhri', 'zarka',
       'lixue', 'rovzan', 'jakob', 'bernd', 'berta', 'armina'],
      dtype=object)

In [86]:
treatments_cut['surname'].unique()

array(['resanovic', 'heilmann', 'svensson', 'luong', 'ribeiro', 'sjoberg',
       'miles', 'adamski', 'citizen', 'lavrentyev', 'villalpando',
       'michalcikova', 'vaara', 'cindric', 'higa', 'antoun', 'guerra',
       'barlow', 'fiorentino', 'aranda', 'czerwinska', 'villadsen',
       'wozniak', 'wisniewski', 'hrafnsdottir', 'koldenhof', 'fodor',
       'wellish', 'traustadottir', 'pecha', 'izmailov', 'casarez',
       'mustafa', 'brhane', 'shishani', 'jackson', 'okoli', 'hermansen',
       'heimisson', 'dratchev', 'rosario', 'madrid', 'tromp', 'radislav',
       'montagu', 'novosel', 'mueller', 'nucci', 'ekechukwu', 'batukayev',
       'cotton', 'nilsen', 'zsinko', 'fesahaye', 'slezinger', 'adams',
       'jurkovic', 'tucker', 'woodward', 'maslov', 'sultygov', 'lund',
       'fakhoury', 'rap', 'hsueh', 'kishiev', 'jakobsen', 'schneider',
       'napolitani', 'sauve'], dtype=object)

In [87]:
treatments_cut['full_name'] = treatments_cut['given_name'] + ' ' + treatments_cut['surname']
treatments_cut['full_name']

0        jozka resanovic
1     inunnguaq heilmann
2         alwin svensson
3              the luong
4         amanda ribeiro
             ...        
65        rovzan kishiev
66        jakob jakobsen
67       bernd schneider
68      berta napolitani
69          armina sauve
Name: full_name, Length: 70, dtype: object

#### The Baseline and end dose of auralin and novodra are in units and the change in the dose from baseline(start) and end was not mentioned. here i'm' calculating the change for auraline and novodra dose's

In [88]:
# Extracting the start and end doses for Auralin
auralin_start = treatments_cut['auralin'].str.extract(r'(\d+)u -')[0]
auralin_end = treatments_cut['auralin'].str.extract(r'- (\d+)u')[0]

# Converting the extracted values to numeric
treatments_cut['auralin_start'] = pd.to_numeric(auralin_start, errors='coerce')
treatments_cut['auralin_end'] = pd.to_numeric(auralin_end, errors='coerce')

In [89]:
# Extracting the start and end doses for Novodra
novodra_start = treatments_cut['novodra'].str.extract(r'(\d+)u -')[0]
novodra_end = treatments_cut['novodra'].str.extract(r'- (\d+)u')[0]

# Converting the extracted values to numeric
treatments_cut['novodra_start'] = pd.to_numeric(novodra_start, errors='coerce')
treatments_cut['novodra_end'] = pd.to_numeric(novodra_end, errors='coerce')

In [90]:
treatments_cut['start_dose'] = treatments_cut['auralin_start'].combine_first(treatments_cut['novodra_start'])
treatments_cut['end_dose'] = treatments_cut['auralin_end'].combine_first(treatments_cut['novodra_end'])


In [91]:
treatments_cut['drug_type'] = treatments_cut.apply(lambda row: 'Auralin' if pd.notna(row['auralin_start']) else 'Novodra', axis=1)

In [92]:
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jozka  resanovic  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        the      luong          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change           full_name  auralin_start  auralin_end  \
0          0.34     jozka resanovic           22.0         30.0   
1           NaN  inunnguaq heilmann           57.0         67.0   
2           NaN      alwin svensson           36.0         39.0   
3          0.92           the luong            NaN          NaN   
4          0.38      amanda ribeiro           36.0         44.0   

   novodra_start  novodra_end  start_dose  end_dose drug_type  
0            NaN          NaN        22.0      30.0   Auralin  
1            NaN          NaN        57.0      67.0   Auralin  
2            NaN          NaN        36.0      39.0   Auralin  
3           61.0         64.0        61.0      64.0   Novodra  
4            NaN          NaN        36.0      44.0   Auralin

In [93]:
treatments_cut.drop(['auralin','novodra'], axis = 1 ,inplace = True)

In [94]:
treatments_cut['hba1c_change'] = treatments_cut['hba1c_start'] - treatments_cut['hba1c_end']

In [95]:
treatments_cut.head()

given_name    surname  hba1c_start  hba1c_end  hba1c_change  \
0      jozka  resanovic         7.56       7.22          0.34   
1  inunnguaq   heilmann         7.85       7.45          0.40   
2      alwin   svensson         7.78       7.34          0.44   
3        the      luong         7.64       7.22          0.42   
4     amanda    ribeiro         7.85       7.47          0.38   

            full_name  auralin_start  auralin_end  novodra_start  novodra_end  \
0     jozka resanovic           22.0         30.0            NaN          NaN   
1  inunnguaq heilmann           57.0         67.0            NaN          NaN   
2      alwin svensson           36.0         39.0            NaN          NaN   
3           the luong            NaN          NaN           61.0         64.0   
4      amanda ribeiro           36.0         44.0            NaN          NaN   

   start_dose  end_dose drug_type  
0        22.0      30.0   Auralin  
1        57.0      67.0   Auralin  
2        36.0      39.0   Auralin  
3        61.0      64.0   Novodra  
4        36.0      44.0   Auralin

In [96]:
column_order = ['given_name', 'surname', 'full_name', 'auralin_start', 'auralin_end', 'novodra_start', 'novodra_end',
       'start_dose', 'end_dose',  'hba1c_start', 'hba1c_end', 'hba1c_change','drug_type']
treatments_cut = treatments_cut[column_order]

In [97]:
treatments_cut.head()

given_name    surname           full_name  auralin_start  auralin_end  \
0      jozka  resanovic     jozka resanovic           22.0         30.0   
1  inunnguaq   heilmann  inunnguaq heilmann           57.0         67.0   
2      alwin   svensson      alwin svensson           36.0         39.0   
3        the      luong           the luong            NaN          NaN   
4     amanda    ribeiro      amanda ribeiro           36.0         44.0   

   novodra_start  novodra_end  start_dose  end_dose  hba1c_start  hba1c_end  \
0            NaN          NaN        22.0      30.0         7.56       7.22   
1            NaN          NaN        57.0      67.0         7.85       7.45   
2            NaN          NaN        36.0      39.0         7.78       7.34   
3           61.0         64.0        61.0      64.0         7.64       7.22   
4            NaN          NaN        36.0      44.0         7.85       7.47   

   hba1c_change drug_type  
0          0.34   Auralin  
1          0.40   Auralin  
2          0.44   Auralin  
3          0.42   Novodra  
4          0.38   Auralin

In [98]:
treatments_cut.head()

given_name    surname           full_name  auralin_start  auralin_end  \
0      jozka  resanovic     jozka resanovic           22.0         30.0   
1  inunnguaq   heilmann  inunnguaq heilmann           57.0         67.0   
2      alwin   svensson      alwin svensson           36.0         39.0   
3        the      luong           the luong            NaN          NaN   
4     amanda    ribeiro      amanda ribeiro           36.0         44.0   

   novodra_start  novodra_end  start_dose  end_dose  hba1c_start  hba1c_end  \
0            NaN          NaN        22.0      30.0         7.56       7.22   
1            NaN          NaN        57.0      67.0         7.85       7.45   
2            NaN          NaN        36.0      39.0         7.78       7.34   
3           61.0         64.0        61.0      64.0         7.64       7.22   
4            NaN          NaN        36.0      44.0         7.85       7.47   

   hba1c_change drug_type  
0          0.34   Auralin  
1          0.40   Auralin  
2          0.44   Auralin  
3          0.42   Novodra  
4          0.38   Auralin

In [99]:
treatments_cut.drop(['auralin_start', 'auralin_end', 'novodra_start', 'novodra_end'], axis = 1, inplace = True)

In [100]:
treatments_cut.to_csv("treatments_cut_cleaned.csv",index = False)

In [101]:
df = pd.read_csv("/content/treatments_cut_cleaned.csv")
df.head()

given_name    surname           full_name  start_dose  end_dose  \
0      jozka  resanovic     jozka resanovic        22.0      30.0   
1  inunnguaq   heilmann  inunnguaq heilmann        57.0      67.0   
2      alwin   svensson      alwin svensson        36.0      39.0   
3        the      luong           the luong        61.0      64.0   
4     amanda    ribeiro      amanda ribeiro        36.0      44.0   

   hba1c_start  hba1c_end  hba1c_change drug_type  
0         7.56       7.22          0.34   Auralin  
1         7.85       7.45          0.40   Auralin  
2         7.78       7.34          0.44   Auralin  
3         7.64       7.22          0.42   Novodra  
4         7.85       7.47          0.38   Auralin

In [102]:
df1 = pd.read_csv("/content/treatments_cleaned.csv")
df1.head()

given_name     surname          full_name  start_dose  end_dose  \
0   veronika    jindrova  veronika jindrova        41.0      48.0   
1     elliot  richardson  elliot richardson        40.0      45.0   
2   yukitaka    takenaka  yukitaka takenaka        39.0      36.0   
3       skye  gormanston    skye gormanston        33.0      36.0   
4     alissa      montez      alissa montez        33.0      29.0   

   hba1c_start  hba1c_end  hba1c_change drug_type  
0         7.63       7.20          0.43   Auralin  
1         7.56       7.09          0.47   Novodra  
2         7.68       7.25          0.43   Novodra  
3         7.97       7.62          0.35   Auralin  
4         7.78       7.46          0.32   Novodra

In [103]:
df2 = pd.read_csv("/content/patients_cleaned.csv")
df2.head()

patient_id given_name  surname    full_name assigned_sex      weight  \
0           1        Zoe  Wellish  Zoe Wellish       female   55.202146   
1           2     Pamela     Hill  Pamela Hill       female   53.886730   
2           3        Jae   Debord   Jae Debord         male   80.648658   
3           4       Liem     Phan    Liem Phan         male  100.198473   
4           5        Tim  Neudorf  Tim Neudorf         male   87.225742   

   height   birthdate  age  phone_number                      email  \
0  1.6764  1976-07-10   43    9517199170   ZoeWellish@superrito.com   
1  1.6764  1967-04-03   52    2175693204       PamelaSHill@cuvox.de   
2  1.8034  1980-02-19   39    4023636804       JaeMDebord@gustr.com   
3  1.7780  1951-07-26   68    7326368246  PhanBaLiem@jourrapide.com   
4  0.6858  1928-02-18   91    3345157487        TimNeudorf@cuvox.de   

                     address               city       state  zip_code  \
0       576 Brown Bear Drive  Rancho California  California     92390   
1  2370 University Hill Road          Armstrong    Illinois     61812   
2      1493 Poling Farm Road               York    Nebraska     68467   
3        2335 Webster Street         Woodbridge  New Jersey      7095   
4       1428 Turkey Pen Lane             Dothan     Alabama     36303   

         country         bmi      bmi_status  
0  United States   19.642676   Normal weight  
1  United States   19.174609   Normal weight  
2  United States   24.797792   Normal weight  
3  United States   31.695499         Obesity  
4  United States  185.459767  Severe Obesity

In [104]:
df3 = pd.read_csv("/content/adverse_reactions_cleaned.csv")
df3.head()

given_name     surname          full_name           adverse_reaction
0      berta  napolitani   berta napolitani  injection site discomfort
1       lena        baer          lena baer               hypoglycemia
2     joseph         day         joseph day               hypoglycemia
3     flavia  fiorentino  flavia fiorentino                      cough
4    manouck     wubbels    manouck wubbels          throat irritation